content/courses/advanced-ada/chapters/types.rst

```diff
@@ -1239,15 +1239,240 @@ Valid
+    end Show_Unchecked_Union;
+
+In the example above, we're use an aggregate in the assignments to :ada:`V`. By
+doing so, we avoid that :ada:`Use_Enum` is set to the *wrong* component. For
+example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Unchecked_State_Or_Integer
+
+    with States; use States;
+
+    procedure Show_Unchecked_Union is
+       V : Unchecked_State_Or_Integer (Use_Enum => True);
+    begin
+       V.S := On;
+       Display_State_Value (V);
+
+       V.I := 2;  --  Error: cannot directly assign to V.I, as Use_Enum is
+                  --         set to True.
+       Display_State_Value (V);
+    end Show_Unchecked_Union;
+
+Here, even though the record has the :ada:`Unchecked_Union` attribute, we
+cannot directly assign to the :ada:`I` component because :ada:`Use_Enum` is set
+to :ada:`True`, so only the :ada:`S` is accessible. We can, however, read its
+value, as we do in the :ada:`Display_State_Value` procedure.
+
+Be aware that, due to the fact the union is not checked, we might write invalid
+data to the record. In the example below, we initialize the :ada:`I` component
+with 10, which is a valid integer value, but results in an invalid value for
+the :ada:`S` component, as the value 10 cannot be mapped to the representation
+of the :ada:`State` type.
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Unchecked_State_Or_Integer
+    :class: ada-run-expect-failure
+
+    with States; use States;
+
+    procedure Show_Unchecked_Union is
+       V : Unchecked_State_Or_Integer (Use_Enum => True);
+    begin
+       V := (Use_Enum => False, I => 10);
+       Display_State_Value (V);
+    end Show_Unchecked_Union;
+
+To mitigate this problem, we could use the :ada:`Valid` attribute |mdash|
+discussed in the previous section |mdash| for the :ada:`S` component before
+trying to use its value in the implementation of the :ada:`Display_State_Value`
+procedure:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Unchecked_State_Or_Integer
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body States is
+
+       procedure Display_State_Value (V : Unchecked_State_Or_Integer) is
+       begin
+          if V.S'Valid then
+             Put_Line ("State: " & V.S'Image);
+          else
+             Put_Line ("State: <invalid>");
+          end if;
+          Put_Line ("Value: " & V.I'Image);
+       end Display_State_Value;
+
+    end States;
+
+    with States; use States;
+
+    procedure Show_Unchecked_Union is
+       V : Unchecked_State_Or_Integer (Use_Enum => True);
+    begin
+       V := (Use_Enum => False, I => 10);
+       Display_State_Value (V);
+    end Show_Unchecked_Union;
 
-    Complete section!
 
+However, in general, you should avoid using the :ada:`Unchecked_Union` aspect
+due to the potential issues you might introduce into your application. In most
+cases, unless you're doing low-level programming or interfacing with C code,
+you won't really need this aspect.
```

Comment:
So much so that I don't think we should be telling them about it at all...

And I disagree that it is needed in low-level programming, the svd2ada generator notwithstanding. :-)

Reply:
I see what you mean, and I do agree that this topic is much less important than other topics I have on my list. 

My goal, however, is to provide a simple and short explanation about aspects that an Ada developer might encounter when reading someone else's code. Developers should be able to get a rudimentary understanding of what unchecked unions are by consulting this "Advanced Ada" course, even if they're not supposed / expected to use them. Unfortunately, reading the RM-B-3-3 section isn't an option in this case, as the text is too technical there.